GoMeet/src/test/java/com/google/sps/EmailServletTest.java

```diff
@@ -0,0 +1,108 @@
+  private final String GUEST_1 = "guest1@gmail.com";
+  private final String GUEST_2 = "guest2@gmail.com";
+  private final String GUEST_3 = "guest3@gmail.com";
+  private final String GUEST_LIST_URI = "guest1%40gmail.com%2Cguest2%40gmail.com%2Cguest3%40gmail.com"; 
+
+  // Mocks
+  private HttpServletRequest mockedRequest;
+  private HttpServletResponse mockedResponse;
+  private EmailServlet mockedServlet;
+  private StringWriter stringWriter;
+  private PrintWriter writer;
+
+  @Before
+  public void setUp() throws IOException {
+    mockedRequest = mock(HttpServletRequest.class);      
+    mockedResponse = mock(HttpServletResponse.class);
+    stringWriter = new StringWriter();
+    writer = new PrintWriter(stringWriter);
+    when(mockedResponse.getWriter()).thenReturn(writer);
+    when(mockedRequest.getParameter("meetingEventId")).thenReturn(MEETING_EVENT_ID);
+    when(mockedRequest.getParameter("guestList")).thenReturn(GUEST_LIST_URI);  
+  }
+
+  @Test
+  public void testSuccessfulDoPost() throws IOException {
+    try (MockedStatic<EmailServlet> mockedServlet = Mockito.mockStatic(EmailServlet.class)) {
+      mockedServlet.when(() -> EmailServlet.sendEmail(GUEST_1, MEETING_EVENT_ID)).thenReturn(true);
+      mockedServlet.when(() -> EmailServlet.sendEmail(GUEST_2, MEETING_EVENT_ID)).thenReturn(true);
+      mockedServlet.when(() -> EmailServlet.sendEmail(GUEST_3, MEETING_EVENT_ID)).thenReturn(true);
+      new EmailServlet().doPost(mockedRequest, mockedResponse);
+
+      // Check the sendEmail function was called exactly once for each guest in the guest list 
+      mockedServlet.verify(times(1), () -> EmailServlet.sendEmail(GUEST_1, MEETING_EVENT_ID));
+      mockedServlet.verify(times(1), () -> EmailServlet.sendEmail(GUEST_2, MEETING_EVENT_ID));
+      mockedServlet.verify(times(1), () -> EmailServlet.sendEmail(GUEST_3, MEETING_EVENT_ID));
+    }
+
+    // Assert that key is returned in JSON Object
+    HashMap<String, Object> keyObj = new HashMap<String, Object>() {{
+      put("sent", true);
+    }};
+    writer.flush(); // writer may not have been flushed yet
+    assertTrue(stringWriter.toString().contains(ServletUtil.convertMapToJson(keyObj)));
+  }
+
+  @Test 
+  public void testNullMeetingEventId() throws IOException {
+    when(mockedRequest.getParameter("meetingEventId")).thenReturn(null);
+    new EmailServlet().doPost(mockedRequest, mockedResponse); 
+    testBadRequest(HttpServletResponse.SC_BAD_REQUEST, ErrorMessages.BAD_EMAIL_REQUEST_ERROR);
+  }
+
+  @Test 
+  public void testNullGuestList() throws IOException {
+    when(mockedRequest.getParameter("guestList")).thenReturn(null); 
+    new EmailServlet().doPost(mockedRequest, mockedResponse); 
+    testBadRequest(HttpServletResponse.SC_BAD_REQUEST, ErrorMessages.BAD_EMAIL_REQUEST_ERROR);
+  }
+
+  private void testBadRequest(int status, String errorMessage) throws IOException {
+    // Check that error response was sent with appropriate details
```

Comment:
As you mentioned in our meeting, this is useful general function that maybe can be move to `ServletUtil`, because I can refactor my tests to use it too 😊
But the method signature would need to be changed to:
`private void testBadRequest(int status, String errorMessage, HttpServletResponse mockedResponse, StringWriter stringWriter)` 
- do you think this parameter list is too long?

Reply:
Yeh that's what I was unsure about too. But I think it makes more sense to have a long parameter list than to repeat the same helper method in all of our test files. I'll see what the hosts says when I asked them to review it after you and Therese have had a look. :)

Reply:
Also not sure if it should be included in ServletUtil or a separate class called ServletTestUtil

Reply:
I think it would be better to include in the test folder, so ServletTestUtil :))

Reply:
Hmm that is true... I think a new class ServletTestUtil, since it has nothing to do with actual servlet functionality.

Reply:
Reusing code and avoiding duplication is great, if you think this can be reused in other tests, then it makes sense to put it somewhere where the other tests can access it.

Reply:
Yep, this function has been added to a ServletTestUtil class in [PR#40](https://github.com/googleinterns/step271-2020/pull/40) for Anna and Therese to use too :)